stream-chat-android-state/src/main/java/io/getstream/chat/android/offline/plugin/logic/channel/internal/TypingEventPruner.kt

```diff
@@ -0,0 +1,246 @@
+     * if such exists.
+     *
+     *
+     * @param userId The ID of the user tied to the typing event.
+     * @param typingStartEvent The typing event will be added to [typingEvents].
+     */
+    private fun addTypingEvent(userId: String, typingStartEvent: TypingStartEvent) {
+        // Create a new self stopping event
+        val timedTypingStartEvent = TimedTypingStartEvent(
+            coroutineScope = coroutineScope,
+            typingStartEvent = typingStartEvent,
+            userId = userId,
+            delayTimeMs = DEFAULT_DELAY_TIME_MS,
+            removeTypingEvent = {
+                removeTypingEvent(it)
+            }
+        )
+
+        // Cancel the self stopping event you are replacing if one exists
+        typingEvents.getOrDefault(userId, null)?.cancelJob()
+
+        // Replace the old self stopping event and call
+        // the updated typing events listener
+        typingEvents[userId] = timedTypingStartEvent
+        onUpdated(getRawTyping(), getTypingEvent())
+    }
+
+    /**
+     * Removes a typing event with the given key (userId)
+     * if such exists in [typingEvents].
+     *
+     * @param userId The ID of the user tied to the typing event.
+     */
+    private fun removeTypingEvent(userId: String) {
+        typingEvents.getOrDefault(userId, null)?.cancelJob()
+
+        typingEvents.remove(userId)
+        onUpdated(getRawTyping(), getTypingEvent())
+    }
+
+    /**
+     * Produces an up to date raw typing map.
+     *
+     * Used to set a value to
+     * [io.getstream.chat.android.offline.plugin.state.channel.internal.ChannelMutableState.rawTyping].
+     */
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    fun getRawTyping(): Map<String, TypingStartEvent> = typingEvents.mapValues { it.value.typingStartEvent }
+
+    /**
+     * Produces an up to date [TypingEvent] instance.
+     *
+     * Used to set a value to
+     * [io.getstream.chat.android.offline.plugin.state.channel.internal.ChannelMutableState.typing].
+     */
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    fun getTypingEvent(): TypingEvent = typingEvents.values
+        .map { it.typingStartEvent }
+        .sortedBy { typingStartEvent -> typingStartEvent.createdAt }
+        .map { typingStartEvent -> typingStartEvent.user }
+        .let { sortedUsers -> TypingEvent(channelId = channelId, users = sortedUsers) }
+
+    /**
+     * Sets the listener used for typing updates.
+     *
+     * @param onUpdated The lambda that will be called whenever the
+     * list of typing events is updated.
+     */
+    fun setOnTypingEventsUpdatedListener(
+        onUpdated: (
+            rawTypingEvents: Map<String, TypingStartEvent>,
+            typingEvent: TypingEvent,
+        ) -> Unit,
+    ) {
+        this.onUpdated = onUpdated
+    }
+
+    /**
+     * Clears all existing typing updates and posts
+     * the resulting updated lists.
+     */
+    fun clear() {
```

Comment:
Not used anywhere other than in a test currently, but it will be useful when https://github.com/GetStream/stream-chat-android/issues/3931 is implemented.